### PR TITLE
[VTK] VTK Exprtk cmake patch

### PR DESCRIPTION
**Pull request by ArashPartow**
* [x]  Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
* [x]  SHA512s are updated for each updated download
* [x]  The "supports" clause reflects platforms that may be fixed by this new version
* [x]  The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
* [x]  Only one version is added to each modified port's versions file.

----------------------------

This patch is intended to unblock the following PR: **https://github.com/microsoft/vcpkg/pull/29665**

and is based on the following VTK MR: **https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9981**

### CODE REVIEW

**Comment by ArashPartow** [comment]
@ LilyWangLL / @JonLiu1993 / @FrankXie05 when you have a moment can you please review and merge this PR.

**Comment by Cheney-W** [comment]
@ArashPartow Although you added a new patch `FindExprTk.patch`, you did not call it in `portfile.cmake` file.

**Comment by Cheney-W** [comment]
```
-- Applying patch FindExprTk.patch
CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:34 (message):
  Applying patch failed: error: corrupt patch at line 55
```

**Comment by ArashPartow** [comment]
> ```
> -- Applying patch FindExprTk.patch
> CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:34 (message):
>   Applying patch failed: error: corrupt patch at line 55
> ```

@Cheney-W  Is there a way I can test it locally?  as this is the patch currently being used by VTK in their mainline https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9981

**Comment by ArashPartow** [comment]
@Cheney-W  seems like dbow2 is failing: 

```
2023-03-02T08:16:40.4892133Z CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
2023-03-02T08:16:40.4896910Z     Command failed: /mnt/vcpkg-ci/downloads/tools/cmake-3.25.1-linux/cmake-3.25.1-linux-x86_64/bin/cmake /mnt/vcpkg-ci/buildtrees/dbow2/src/cdcda8b401-bf5ac0a10a.clean -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/mnt/vcpkg-ci/packages/dbow2_x64-linux/debug -DBUILD_Demo=OFF -DCMAKE_MAKE_PROGRAM=/mnt/vcpkg-ci/downloads/tools/ninja-1.10.2-linux/ninja -DCMAKE_SYSTEM_NAME=Linux -DBUILD_SHARED_LIBS=OFF -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=/agent/_work/1/s/scripts/toolchains/linux.cmake -DVCPKG_TARGET_TRIPLET=x64-linux -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=external -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=/agent/_work/1/s/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=/agent/_work/1/s -DZ_VCPKG_ROOT_DIR=/agent/_work/1/s -D_VCPKG_INSTALLED_DIR=/mnt/vcpkg-ci/installed -DVCPKG_MANIFEST_INSTALL=OFF
2023-03-02T08:16:40.4899747Z     Working Directory: /mnt/vcpkg-ci/buildtrees/dbow2/x64-linux-dbg
2023-03-02T08:16:40.4900317Z     Error code: 1
2023-03-02T08:16:40.4900596Z     See logs for more information:
2023-03-02T08:16:40.4901111Z       /mnt/vcpkg-ci/buildtrees/dbow2/config-x64-linux-dbg-CMakeCache.txt.log
2023-03-02T08:16:40.4901647Z       /mnt/vcpkg-ci/buildtrees/dbow2/config-x64-linux-dbg-out.log
2023-03-02T08:16:40.4902153Z       /mnt/vcpkg-ci/buildtrees/dbow2/config-x64-linux-dbg-err.log
2023-03-02T08:16:40.4902379Z 
2023-03-02T08:16:40.4902612Z Call Stack (most recent call first):
2023-03-02T08:16:40.4902994Z   scripts/cmake/vcpkg_configure_cmake.cmake:326 (vcpkg_execute_required_process)
2023-03-02T08:16:40.4903520Z   ports/dbow2/portfile.cmake:11 (vcpkg_configure_cmake)
2023-03-02T08:16:40.4907308Z   scripts/ports.cmake:147 (include)
2023-03-02T08:16:40.4907871Z 
2023-03-02T08:16:40.4908022Z 
2023-03-02T08:16:40.4923332Z error: building dbow2:x64-linux failed with: BUILD_FAILED
```

**Comment by Cheney-W** [comment]
```
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/vtk/FindExprTk.cmake:27 (string):
  Syntax error in cmake code at

    /mnt/vcpkg-ci/installed/x64-linux/share/vtk/FindExprTk.cmake:27

  when parsing string

    version = \"(2\.7[0-9.]+)\".*$

  Invalid escape sequence \.
```
Error logs are here: https://dev.azure.com/vcpkg/public/_build/results?buildId=86201&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by ArashPartow** [comment]
@Cheney-W looks good now, can we now please merge this PR.

**Review by Cheney-W — APPROVED** [review]

**Comment by ArashPartow** [comment]
@Cheney-W / @Adela0814  can this please be merged.

**Comment by Cheney-W** [comment]
I have added the info:reviewed label, now this PR is waiting for the developer to merge.

**Comment by ArashPartow** [comment]
> I have added the info:reviewed label, now this PR is waiting for the developer to merge.

thank-you 😃

File: ports/vtk/FindExprTk.patch [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ if (ExprTk_INCLUDE_DIR)
+-  file(STRINGS "${ExprTk_INCLUDE_DIR}/exprtk.hpp" _exprtk_version_header
+-    REGEX "static const char\\* version")
+-  string(REGEX MATCH "static const char\\* version = \"([0-9.]+)\"" _exprtk_version_match "${_exprtk_version_header}")
+-  set(ExprTk_VERSION "${CMAKE_MATCH_1}")
++  file(STRINGS "${ExprTk_INCLUDE_DIR}/exprtk.hpp" _exprtk_version_header REGEX "\"[0-9.]+\"")
++  set(ExprTk_VERSION)
++  foreach (_exprtk_version_line IN LISTS _exprtk_version_header)
++    if ("${ExprTk_VERSION}" STREQUAL "")
++      string(REGEX MATCH "version = \"(2.7[0-9.]+)\".*$" _exprtk_version_match "${_exprtk_version_line}")
```
Comment (dg0yt) [inline ports/vtk/FindExprTk.patch:37]: `.` matches *any*  character. The correct escaping is `\\.`. But I usually prefer `[.]`. This is also used here after `0-9`.

Comment (ArashPartow) [reply ports/vtk/FindExprTk.patch:37]: The sequence \\. will escape \ and then assume . which will mean literally \. there is no backslash in the version.

**Comment by ArashPartow** [comment]
@Adela0814 and @BillyONeal we're awaiting someone with merge permissions to merge this PR.  

atm this PR is blocking the the following: https://github.com/microsoft/vcpkg/pull/29665

File: ports/vtk/FindExprTk.patch (hunk 1 above)
Comment (dg0yt) [reply ports/vtk/FindExprTk.patch:37]: Note that the `\` inside double quotes will escape the next character with regard to double quotes. That's why it takes `\\.` to pass a literal `\.` into a regex.
[CMake regex doc](https://cmake.org/cmake/help/v3.20/command/string.html?highlight=backslashes#regex-replace): "Note that two backslashes (`\\1`) are required in CMake code to get a backslash through argument parsing."

**Comment by dg0yt** [comment]
I wish GH would let me mark [my own comments](https://github.com/microsoft/vcpkg/pull/29955#discussion_r1124155259) as "not resolved".

**Pull request closed without merging**